Add incomplete events

benjben opened the pull request:
This PR adds incomplete events to Enrich flow.

Before, enriching an event [was resulting](https://github.com/snowplow/enrich/blob/4.0.0/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L69) in either a `BadRow` or an `EnrichedEvent`. What's new is that now a `BadRow` [can have](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L69) an `EnrichedEvent` associated with it, which is the incomplete event.

All the changes are related to the fact that now we don't short circuit the processing as soon as there is an error but we keep going with what is valid.

As a result we're not using `ValidatedNel` and `EitherT` any more.

To make things easier to follow, I reorganized the flow in 3 steps:
- [Mapping and validating the input](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L71).
- [Running the enrichments](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L73). This step now takes care of validating the enrichments contexts.
- [Validating the atomic fields lengths](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L89).

istreeter commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/package.scala:
```diff
@@ -25,8 +25,8 @@ package object fs2 {
 
-  /** Enrichment result, containing list of (valid and invalid) results as well as the collector timestamp */
-  type Result = (List[Validated[BadRow, EnrichedEvent]], Option[Long])
+  type Enriched = Either[(BadRow, Option[EnrichedEvent]), EnrichedEvent]
```
So this can be either just a BadRow or a just an EnrichedEvent or both.  Have you ever come across [cats Ior](https://typelevel.org/cats/datatypes/ior.html)?

benjben replied:
I didn't know about `Ior`. It seems like a good fit indeed. But what I don't like is that in case of an incomplete event, it is at the exact same place as a valid enriched event. I kind of like that the incomplete event is in the `Left`, coupled with the `BadRow`. So I'd rather leave it like that, except if you strongly think that we should swich to `Ior`.

istreeter commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -168,26 +163,31 @@ object IgluUtils {
+      .map(_.separate)
+      .map {
+        case (errors, valid) =>
+          val maybeBadRow = errors match {
```
Would be neater and a tiny bit more efficient to do it like:

```scala
val maybeBadRow = errors.toNel.map { nel =>
  val enrichmentFailures = nel.map {
    case (schemaKey, clientError) => ???
  }
  EnrichmentManager.buildEnrichmentFailuresBadRow(.....)
}
```

istreeter commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -48,44 +48,45 @@ object IgluUtils {
+                          case Left(err) => List(err)
+                          case _ => Nil
+                        }
+      maybeBadRow = (invalidContexts ::: invalidUnstruct) match {
```
Consider:

```scala
maybeBadRow = (invalidContexts ::: invalidUnstruct).toNel.map { nel =>
  buildSchemaViolationsBadRow(nel, ???, ???, ???, ???)
}
```

istreeter reviewed (commented):
No real strong objections from me on this.  I do think it has made the code harder to read compared with how it was before, but maybe that's inevitable when adding a complex feature.

The point about `Ior` is that it has a Monad on the right hand side.  So it might allow arranging the code like this:

```scala
for {
  a <- doSomethingThatMightYieldBadRows(collectorPayload)
  b <- doSomethingElseThatMightYieldBadRows(a)
  event <- andSomethingElse(b)
} yield event
```

...so you accumulate failures on the left hand side, while continuing to process the event on the right hand side.

There's even a `IorT` which is similar to an `EitherT`.

I understand this work was done to unblock the next phases of this project, so I don't think we should stick around for too long searching for perfection.

benjben commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:
```diff
@@ -65,44 +64,103 @@ object EnrichmentManager {
-    atomicFields: AtomicFields
-  ): EitherT[F, BadRow, EnrichedEvent] =
+    atomicFields: AtomicFields,
+    emitIncomplete: Boolean = true
```
That's exactly the plan! There is another ticket about adding configuration for the feature

benjben commented:
Thanks for the review @istreeter ! I will address your feedback.

> I do think it has made the code harder to read compared with how it was before

Yeah I feel the same :smiling_face_with_tear: But the logic has become more complicated, so I'm afraid it's an inevitable outcome. Now after each step we need to check if there are bad rows and the value of `emitIncomplete`.

> The point about Ior is that it has a Monad on the right hand side

I'm not sure that we will benefit from it, as we need to check for bad rows after each step, and not only at the end. But I'll try `Ior` to see if it can simplify the workflow.

benjben commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -248,13 +248,14 @@ object IgluUtils {
-        .map(_.sequence_.toEither)
-    }
+  ): F[List[Either[(SchemaKey, ClientError), SelfDescribingData[Json]]]] =
+    sdjs
```
Way neater indeed!

benjben commented:
Hi @istreeter , I addressed your feedback, please have a look. I agree that having `Ior[BadRow, EnrichedEvent]` as the return type of `enrichEvent` looks nicer. Then I tried to use `Ior` everywhere in the flow but I didn't like it. One problem is that to get the errors you need to pattern match and to have `case Ior.Left` whereas we know already that we will *never* be in this case.

> for {
>  a <- doSomethingThatMightYieldBadRows(collectorPayload)
>  b <- doSomethingElseThatMightYieldBadRows(a)
>  event <- andSomethingElse(b)
> } yield event
> ...so you accumulate failures on the left hand side, while continuing to process the event on the right hand side.

I might be missing something because I don't see how this can work. Before executing `doSomethingElseThatMightYieldBadRows` we need to check the errors that are in `a` (not everyone will have incomplete events activated), how would that work? I want to be able to accumulate the errors *and* to continue the processing (so no short circuiting with `IorT`).

An alternative that I envisaged was to use `IorT` and to have a `Ref[F, List[BadRow]]` . We would pass it to each function, along with `emitIncomplete`. Then inside each function, when there is an error, if `emitIncomplete` is set to `false`, we return a `Left` which short circuis the rest of the processing. If set to `true`, we add the bad row to the `Ref` and we return `Right`, which will continue the processing. And at the end we look at what is inside the `Ref` to decide what to return. Do you think that it would be nicer? We can jump on a call if it's unclear.

benjben commented:
Hey @istreeter 

> Imagine for a second that everyone has incomplete events activated. Would that remove the need to check errors after each step, and therefore remove the part that makes it messy?

Indeed that would remove the need for that and would make the flow look nicer.

> Under what circumstances would someone have this feature disabled?

This feature will cost money as it will require an additional stream and additional apps to run, so I can imagine customers not wanting it.

@colmsnowplow should we consider that once this feature exists we will activate it for all our customers ?

> And in those cases, could we continue to process the event, but simply not emit it in the final step?

For sure we can but that would waste resources. That's not very satisfying to force features only to make the code look nicer.

colmsnowplow commented:
I think I can offer a point of view here.

First just to double check I'm reading it right:

> Before executing doSomethingElseThatMightYieldBadRows we need to check the errors that are in a (not everyone will have incomplete events activated), how would that work?

I'm not 100% clear why we need to check the errors, but it sounds like it's so that we can decide not to continue processing the event if we hit an error. Just stating this in case it's not correct. :) 

> Under what circumstances would someone have this feature disabled? 

This could happen! An example - one of the customers in the research uses the JS enrichment for custom bot filtering. They specifically don't want that data to go to the db. First version of this they will most likely want to just disable it. Some customers might simply not want the additional cost of an extra stream and an extra loader.

> And in those cases, could we continue to process the event, but simply not emit it in the final step?

This would be acceptable. But perhaps we can find something cleaner. (Not sure as haven't yet carved out the time to read this PR properly)

colmsnowplow commented:
> @colmsnowplow should we consider that once this feature exists we will activate it for all our customers ?

I think we were editing at the same time, think I answered that at least at first I don't think we can. @stanch, FYI, think you'll agree. :)

benjben commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -122,35 +105,30 @@ object IgluUtils {
-        } yield contextsSDJ
+          contextsSdj <- contexts.traverse(parseAndValidateSDJ(_, client, registryLookup))
+        } yield contextsSdj
+        iorT.recoverWith { case errors => IorT.fromIor[F](Ior.Both(errors, Nil)) }
```
Hey @istreeter , please note that I introduced some `recoverWith`s, as we don't want any `Left`, as it would short circuit the processing, even though incomplete events are enabled

istreeter commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -310,16 +284,8 @@ object IgluUtils {
-      Failure.SchemaViolations(Instant.now(), vs),
-      Payload.EnrichmentPayload(pee, re)
-    )
+  private implicit class ExtraEitherTSyntax[F[_], A, B](val eitherT: EitherT[F, A, B]) extends AnyVal {
+    def toIorT(implicit F: Monad[F]): IorT[F, A, B] =
+      IorT(eitherT.value.map(_.toIor))
+  }
```
I think you don't need this, although I know copied it from me!  EitherT [already has a method](https://github.com/typelevel/cats/blob/v2.10.0/core/src/main/scala/cats/data/EitherT.scala#L707) called `toIor`.  I missed it earlier because I was expecting it to be called `toIorT`.

istreeter commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -243,37 +212,42 @@ object IgluUtils {
-        .sequence
-        .map(_.sequence_.toEither)
+  ): IorT[F, NonEmptyList[(SchemaKey, ClientError)], List[SelfDescribingData[Json]]] =
+    sdjs.traverse { sdj =>
+      check(client, sdj, registryLookup)
+        .map(_ => sdj)
+        .leftMap(NonEmptyList.one)
+        .toIor
```
I think this is wrong -- and it's my fault because my version was also wrong.

Currently, if `sdjs` contains some good and some bad, then I think this returns a `Left` not a `Both`.  I guess you want it to return a `Both`?

Furthermore... if sdjs contains only bad, then I think it would be more logical if this function can return a `Both(failures, Nil)`.  And if you can get that working, then I think you can remove the `recoverWith`s further up the page.

istreeter reviewed (commented):
There are lots of places where you've changed to emitting a SchemaViolation instead of an EnrichmentFailure.  This definitely needs checking with others!

It is a design decision that affects how users interact with the bad data. It deserves a discussion beyond just me and you in a code review.

istreeter commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:
```diff
@@ -243,37 +216,45 @@ object IgluUtils {
-        .map(_.sequence_.toEither)
-    }
+  ): IorT[F, NonEmptyList[(SchemaKey, ClientError)], List[SelfDescribingData[Json]]] =
+    sdjs
```
Sharing this in case you like it:

```scala
    sdjs.map { sdj =>
      check(client, sdj, registryLookup)
        .map(_ => List(sdj))
        .leftMap(NonEmptyList.one)
        .toIor
        .recoverWith { case errors => IorT.fromIor[F](Ior.Both(errors, Nil)) }
    }.foldA
```

benjben replied:
Yes I like it, it's fewer letters :laughing:

benjben commented on modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala:
```diff
@@ -189,11 +190,19 @@ object Enrich {
     env: Environment[F, A]
   ): F[Unit] = {
-    val (bad, enriched) =
+    //val (bad, enriched, incomplete) =
```
This gets removed [here](https://github.com/snowplow/enrich/pull/874/files#diff-4381e8d9cbe4dd5147c2491673ea2686da148ae9782c0d228e59ea766981e8e4R195).

Now that this PR is approved, and that the other one about the configuration is also approved, I think that I'll just merge it in this one. This PR will be the main PR for incomplete events and all the new stuff will get merged into it as we go. And when we're fully done with the feature we'll merge it to `develop`

colmsnowplow replied:
So if I understand correctly, this branch is the working branch, other work gets based off this one, until we're ready to merge to develop.

Sounds good to me!

benjben replied:
Yes exactly :+1:

benjben commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:
```diff
@@ -83,26 +91,55 @@ object EnrichmentManager {
+             .possiblyExitingEarly(emitIncomplete)
     } yield enriched
 
+    iorT.leftMap(_.last)
```
Turns out that it was `.head` ! Testing FTW

benjben commented on modules/common/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/EnrichmentManagerSpec.scala:
```diff
@@ -167,29 +167,32 @@ class EnrichmentManagerSpec extends Specification with EitherMatchers with CatsE
               ) =>
-            schemaKey1 must beEqualTo(emailSentSchema)
-            clientError1.toString must contain(
+            schemaKey1 must beEqualTo(AtomicFields.atomicSchema)
+            clientError1.toString must contain("tr_tt")
+            clientError1.toString must contain("Cannot be converted to java.math.BigDecimal")
+            schemaKey2 must beEqualTo(emailSentSchema)
+            clientError2.toString must contain(
```
Interestingly the order is not the same when combining several `ValidatedNel[A, B]` and several `Ior[List[A], B]`

spenes commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:
```diff
@@ -68,44 +68,114 @@ object EnrichmentManager {
+                           .leftMap(NonEmptyList.one)
+                           .possiblyExitingEarly(emitIncomplete)
+        // Next 2 lines remove the invalid contexts and the invalid unstructured event from the event.
+        // This should be done after the bad row was created and only if emitIncomplete is enabled.
```
I think this comment should be removed because next two lines should be performed in all cases not just when `emitIncomplete` is enabled because `contexts` and `unstruct_event` might need to be updated if there is self describing json with superseded schema.

colmsnowplow commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:
```diff
@@ -68,44 +68,112 @@ object EnrichmentManager {
+      for {
+        enrichedEvent <- Sync[F].delay(new EnrichedEvent)
+        enrichmentResult <- enrich(enrichedEvent).value
+        now = Instant.now()
+        _ = setDerivedContexts(enrichedEvent, enrichmentResult, now, processor)
```
I don't remember the discussions we had about it early on. 

I think my position on these design aspects was that we should make a best effort for this kind of thing but the implementation difficulty also has a role to play.

If it's not difficult to have it so that the tstamp is different for each, then I would say do it. If it is, then I would say let @stanch worry about whether or not that requirement becomes important enough to spend time on later.

@benjben I would defer to your call on it either way

benjben replied:
I can't find where but it was agreed that we would use the same timestamp everywhere to simplify the implementation and because we're not sure that having different timestamps is important.

istreeter replied:
Do we have any examples yet of how this timestamp will be used at all?  e.g. any recommended sql query that might use this timestamp for recoveries?

Sorry to keep harping on about this timestamp.  If find it interesting from the point of view of schema design -- and how this timestamp complements the existing `etl_tstamp`.

colmsnowplow replied:
[This is the only note I have recorded about it](https://www.notion.so/keep-in-the-snow/Failure-Entity-Schema-Design-6361207b9801468ca8ac2bbf448c81f0?pvs=4#f8d99bec76b44e3383a009409b11c2ff).

I do vaguely remember the convo we had before about it - if the use case of ordering failures isn't doable then I personally don't feel massively strongly about keeping it or dropping it. 

My only note on the topic is that we should do our best to avoid future breaking changes to this scheme where possible - but I don't think this decision could cause that.

^^ That's my 2c on it anyway! @benjben you'll be much more better with this more recently so I would for sure weigh your input more heavily than mine at this point :)

benjben replied:
I'm sure that it was agreed on that it was fine to use the same timestamp everywhere, at least for now, but I can't find where on Slack, Github and Notion.

I'd rather avoid to make last minute changes like that, but let me check if it's quick enough to do.

colmsnowplow replied:
Sorry if my last message was unclear - I meant to communicate that I think that was agreed, but my memory of the details of the conversation is hazy, so just offering thoughts on the topic in general.

Apologies if how I phrased it made it sound like I was suggesting we make a change now - that wasn't the intention!

benjben replied:
No worries at all, I hadn't misinterpreted what you said. But I was still curious about how easy/hard it is to use different timestamps, as we should aim at what the best is. It seems quite straightforward to do, I'm giving it a go.

benjben replied:
I implemented the change so that each failure gets its own timestamp at the time of creation, could you have a look @istreeter please ?

With the latest implementation it was quite straightforward, which was probably not the case when we decided that it was fine to use the same.

benjben commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/AtomicFields.scala:
```diff
@@ -31,97 +33,98 @@ object AtomicFields {
-      AtomicField(name = "event_format",       _.event_format       ),
-      AtomicField(name = "event_version",      _.event_version      ),
-      AtomicField(name = "event_fingerprint",  _.event_fingerprint  ),
+      AtomicField(name = "app_id", _.app_id, _.app_id = null),
```
Same, not sure what happened. Let me put it back as it was

benjben commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala:
```diff
@@ -0,0 +1,191 @@
+        (
+          List(
+            Json.obj(
+              "message" := s"$field - $expectation",
+              "source" := field
```
Yes it's deliberate, but thanks for asking !

istreeter commented on modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala:
```diff
@@ -0,0 +1,191 @@
+        val error = Json.obj("message" := err, "source" := v.source)
+        ("NotJSON", List(error), None, Json.obj(v.source := v.data).some)
+      case FailureDetails.SchemaViolation.NotIglu(_, err) =>
+        val message = err.message("").split(":").headOption
```
Just an idea... I would have more confidence in the maintainability of this code if we changed it to be more explicit.

```scala
val message = err match {
  case ParseError.InvalidSchemaVer => "Invalid schema version"
  case ParseError.InvalidIgluUri => "Invalid Iglu URI"
  case ParseError.InvalidData => "Invalid data payload"
  case ParseError.InvalidSchema => "Invalid schema"
  case ParseError.InvalidMetaschema => "Invalid metashema"
}
```

benjben replied:
Much better indeed! Thank you!

istreeter commented:
I had already approved it, but it looks even better now ✅
